fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -85,122 +42,646 @@ def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+                self.batches[chunk] = Batch(tuple(tensors), chunk)
+
+    def sync_stream(self, chunk: int, stream: torch.cuda.Stream) -> None:
+        """syncs the stream with cuda events associated with transmission of the chunck to the cuda device."""
+        for e in self.recv_events[chunk]:
+            if e is not None:
+                stream.wait_event(e)
+
+    def get_batch(self, chunk: int) -> Batch:
+        batch = self.batches[chunk]
+        assert batch is not None
+        return batch
+
+
+class PartitionHandler:
+    """This class processes a single partition of the pipeline.
+    Args:
+        module_rref: RRef to the nn.Module for this partition. It should be on the local rpc worker.
+        device: The device that holds the module.
+        num_inputs: Numer of inputs to the module
+        num_outputs: Number of outputs of the module. If the module output is not a tuple (and it is a
+            single tensor), num_outputs should be None.
+        rank: The rank of the partition
+        chunks: Number of micor-batches in a mini-batch
+        checkpoint_stop:: Checkpointing is done only for the first checkpoint_stop chunks of a mini-batch.
+    """
 
     def __init__(
         self,
-        module: List[LayerSpec],
-        *,
-        balance: List[int],
-        devices: List[str],
-        chunks: int = chunks,
-        checkpoint: str = checkpoint,
-        deferred_batch_norm: bool = False,
+        module_rref: rpc.RRef,
+        device: str,
+        num_inputs: int,
+        num_outputs: Optional[int],
+        rank: int,
+        chunks: int,
+        checkpoint_stop: int,
     ) -> None:
+        self.module = module_rref.local_value()
+        self.chunks = chunks
+        self.device = torch.device(device)
+        self.checkpoint_stop = checkpoint_stop
+        self.rank = rank
+        self.num_inputs = num_inputs
+        self.num_outputs = num_outputs
+        (self.in_queue,), (self.out_queue,) = create_workers([self.device])
+
+    def local_parameter_rrefs(self) -> List[rpc.RRef]:
+        r"""
+        Create one RRef for each parameter in the given local module, and return a
+        list of RRefs.
+        """
+        return [rpc.RRef(p) for p in self.module.parameters()]
+
+    def make_pipeline_record(self, consumers: List[OutputConsumer]) -> DistributedPipelineRecord:
+        return DistributedPipelineRecord(
+            self.device, self.rank, self.chunks, self.num_inputs, self.num_outputs, consumers
+        )
+
+    def run(self, pipeline_record: DistributedPipelineRecord) -> None:
+        """Runs pipeline parallelism. It modifies the given batches in place."""
+        m = len(pipeline_record.batches)
+
+        self.stream = current_stream(self.device)
+
+        for chunk in range(m):
+            with record_function("feed"):
+                pipeline_record.wait_for(chunk)
+            self.fence(pipeline_record, chunk)
```

Comment:
Should fence be a method on pipeline record. It feels like there are too many internals of pipeline_record being exposed here. Should PartitionHandler and PipelineRecord be one single class? Or should more of the logic in this class be pushed to PipelineRecord so that it is hiding more details and has a simpler interface.